### PR TITLE
[Feature #20005] Add C API to resolve symbols from other extensions

### PR DESCRIPTION
**Pull request by tagomoris**
This is to add a C API for extensions to resolve symbols `extern`-ed by other extensions.

https://bugs.ruby-lang.org/issues/20005

### CODE REVIEW

**Comment by tagomoris** [comment]
I have no idea how we can test this kind of API. What's the best example?

**Comment by tagomoris** [comment]
I got some idea about `ext/-test-` and `test/-ext-`. So I'll try them anyway.

**Comment by KJTsanaktsidis** [comment]
That sounds right - make two modules in `ext/-test-`, and make one try and resolve symbols from the other. I think the  feature name for an extension `foo` in `ext/-test-/foo` will be `-test-/foo`

**Comment by tagomoris** [comment]
I added extensions for tests under `ext/-test-` and also added a test case under `text/-ext-`. These look working correctly on my laptop.

File: load.c [hunk 1]
```diff
@@ -1511,6 +1515,29 @@ rb_f_autoload_p(int argc, VALUE *argv, VALUE obj)
...
+    }
+    if (rb_ary_entry(resolved, 0) != ID2SYM(rb_intern("so"))) {
+        return NULL;
+    }
+    path = rb_ary_entry(resolved, 1);
+    handle = rb_hash_aref(ruby_dln_libmap, path);
+    if (NIL_P(handle)) {
+        return NULL;
+    }
+    return dln_symbol((void *)NUM2LONG(handle), symbol);
```
Comment (nobu) [inline load.c:1538]:
```suggestion
    return dln_symbol((void *)NUM2SVALUE(handle), symbol);
```

File: load.c [hunk 2]
```diff
@@ -1511,6 +1515,29 @@ rb_f_autoload_p(int argc, VALUE *argv, VALUE obj)
...
+
+    resolved = rb_resolve_feature_path((VALUE)NULL, fname_str);
+    if (NIL_P(resolved) || RARRAY_LEN(resolved) != 2) {
+        return NULL;
+    }
+    if (rb_ary_entry(resolved, 0) != ID2SYM(rb_intern("so"))) {
+        return NULL;
+    }
+    path = rb_ary_entry(resolved, 1);
+    handle = rb_hash_aref(ruby_dln_libmap, path);
```
Comment (nobu) [inline load.c:1534]: `rb_hash_aref` unnecessarily uses the default value.
 
```suggestion
    handle = rb_hash_lookup(ruby_dln_libmap, path);
```

File: load.c [hunk 3]
```diff
@@ -1511,6 +1515,29 @@ rb_f_autoload_p(int argc, VALUE *argv, VALUE obj)
...
+{
+    VALUE handle;
+    VALUE resolved;
+    VALUE path;
+    VALUE fname_str = rb_str_new_cstr(fname);
+
+    resolved = rb_resolve_feature_path((VALUE)NULL, fname_str);
+    if (NIL_P(resolved) || RARRAY_LEN(resolved) != 2) {
+        return NULL;
+    }
```
Comment (nobu) [inline load.c:1529]: You might need to consider statically linked extensions here.

**Review by nobu — CHANGES_REQUESTED** [review]

File: dln.c [hunk 4]
```diff
@@ -447,6 +447,33 @@ dln_sym(void *handle, const char *symbol)
...
+#if defined(_WIN32) || defined(USE_DLN_DLOPEN)
+    if (EXTERNAL_PREFIX[0]) {
+        const size_t symlen = strlen(symbol);
+        char *const tmp = ALLOCA_N(char, symlen + sizeof(EXTERNAL_PREFIX));
+        if (!tmp) dln_memerror();
+        memcpy(tmp, EXTERNAL_PREFIX, sizeof(EXTERNAL_PREFIX) - 1);
+        memcpy(tmp + sizeof(EXTERNAL_PREFIX) - 1, symbol, symlen + 1);
+        symbol = tmp;
+    }
+    if (handle == NULL) {
```
Comment (tagomoris) [inline dln.c:462]: The API to pass `NULL` as the `handle` argument looks dirty - should I add another argument to search the main executable explicitly?

File: dln.c [hunk 5]
```diff
@@ -447,6 +447,33 @@ dln_sym(void *handle, const char *symbol)
...
+# if defined(USE_DLN_DLOPEN)
+        handle = dlopen(NULL, RTLD_FIRST);
+# else
+        // WIN32 seems not having similar feature to get the handle of the main process
```
Comment (tagomoris) [inline dln.c:466]: Is this true? I couldn't find any way to do that.

File: include/ruby/internal/intern/load.h [hunk 6]
```diff
@@ -176,6 +176,19 @@ VALUE rb_f_require(VALUE self, VALUE feature);
...
+/**
+ * Resolves and returns a symbol in the native extension specified by the
+ * feature and symbol names. Extensions will use this function to access the
+ * symbols provided by other native extensions.
+ *
+ * @param[in]  feature           Name of a feature, e.g. `"json"`.
+ * @param[in]  symbol            Name of a symbol defined by the feature.
+ * @return     The resolved symbol, defined and externed by the specified
+ *             feature. It may be NULL when the feature is not loaded, the
+ *             feature was not extension, or the symbol was not found.
```
Comment (tagomoris) [inline include/ruby/internal/intern/load.h:188]: Actually, the current implementation raises LoadError when the feature exists but the symbol is not found.
Should it suppress the LoadError?

File: load.c (hunk 3 above)
Comment (tagomoris) [reply load.c]: I updated the code on the assumption that the feature name of statically linked extensions are also included in `$LOADED_FEATURES`. Is that true?

File: include/ruby/internal/intern/load.h (hunk 6 above)
Comment (nobu) [reply include/ruby/internal/intern/load.h:188]: It is questionable to me.

File: dln.c (hunk 5 above)
Comment (nobu) [reply dln.c:466]: `rb_libruby_handle()` should return the handle to `libruby`.
Rather `RTLD_FIRST` seems platform dependent.

File: load.c (hunk 3 above)
Comment (nobu) [reply load.c]: You can use `rb_feature_p()` instead, I guess.

File: include/ruby/internal/intern/load.h (hunk 6 above)
Comment (tagomoris) [reply include/ruby/internal/intern/load.h:188]: Finally, I think that returning NULL in both cases (the feature not found and the symbol not found) should be consistent.

**Comment by tagomoris** [comment]
CIs are failing on Windows environments. These show `rb_ext_resolve_symbol` is failing to resolve symbols. Hmm?

**Comment by tagomoris** [comment]
And YJIT-ed runs crashed.

**Comment by tagomoris** [comment]
About the CIs on Windows environments, the LoadError message `<LoadError: resolve_symbol_target is not loaded: 80>` explains that `GetProcAddress(handle, symbol);` returns NULL with a valid handle (not statically linked).

**Comment by tagomoris** [comment]
I thought it was enough to specify `RUBY_EXTERN` to export a symbol from a .dll, but failing tests show it's not. Does it need to have `*.def` to declare `EXPORTS` explicitly? Or is there anything I'm missing?
I'll investigate more later.

**Comment by tagomoris** [comment]
All tests passed!

**Comment by tagomoris** [comment]
I've updated my commits and pushed again.
@nobu Could you review this change once more?

**Review by nobu — APPROVED** [review]

**Comment by tagomoris** [comment]
Can this get merged into Ruby 3.3?

**Review by nobu — APPROVED** [review]

**Pull request merged**